source/agora/consensus/EnrollmentManager.d

```diff
@@ -565,12 +565,13 @@ public class EnrollmentManager
         foreach (const ref key; keys)
         {
             const preimage = this.validator_set.getPreimageAt(key, height);
-            // this should not happen. validators which didn't reveal the
-            // preimage should not be in the active validator set 'keys'.
+            // We could have a misbehaving validator that does not reveal
```

Comment:
The way I envisioned it is that we would simply return the preimage root that's stored in the block header.

This function is always called with a block height in the past.

Reply:
But I guess the next commits do that. I'm still reviewing.

Reply:
> The way I envisioned it is that we would simply return the preimage root that's stored in the block header.
> 
> This function is always called with a block height in the past.

I didn't catch that before. By starting to add `random_seed` in the header of the `GenesisBlock`, we can use the `random_seed` in the way you envisioned. I will add the code about that in this branch.

Reply:
> The way I envisioned it is that we would simply return the preimage root that's stored in the block header.
> 
> This function is always called with a block height in the past.

I thought about it. 
The `random_seed` of a block header is for the previous height and the `rand_seed` from this `getRandomSeed` function is for the current height. So we should generate a random seed again in this function, I think.

Reply:
> which is used for the next block 

I thought this function was used for the quorum generator. When it generates quorums for height N it uses preimages of N-1. So it could just read the root from the N-1 block header.

Reply:
I think the problem I am seeing here is that `getRandomSeed` is used by both the quorum generator and the slasher.

The quorum generator needs a fixed (externalized) random seed, which we are now storing in the block header.

So maybe there should be two functions? One is the `getCandidateRandomSeed` and the other `getExternalizedRandomSeed`. Do you see what I mean?

Reply:
> > which is used for the next block
> 
> I thought this function was used for the quorum generator. When it generates quorums for height N it uses preimages of N-1. So it could just read the root from the N-1 block header.

I thought it the same way, but I realized that the random seed from the N-1 block header is generated with preimages of N-2 in rethinking. Do you mean that the random seed is generated with pre-images of current validators and newly enrolled validators?

Reply:
> So maybe there should be two functions? One is the `getCandidateRandomSeed` and the other `getExternalizedRandomSeed`. Do you see what I mean?

I got it. I thought it right before. We can generate the random seed for the `getExternalizedRandomSeed` with pre-images of current validators and newly enrolled validators when validating a block, which is stored in a block header after.

Reply:
So I took another look now.

When a block is externalized, we call `rebuildQuorumConfig(.., block.height)`. That function calls `getRandomSeed()`.

What I am worried about is the case where a validator reveals their preimage in the small time window just before a block is externalized. Their preimage might not be included in the block's header (because it's too late),, but it might get included in some node's local validator set because a new block was not externalized yet. Then, when `rebuildQuorumConfig()` is called it's possible some nodes will use a different random seed based on their current internal preimage state.

That's why I suggest to use a separate function for the quorum generator. It should take the random seed in the block header (always). I was wrong about it taking the `N-1`'th preimages, it's the `N`'th preimages.

Does that make sense?

Reply:
> That's why I suggest to use a separate function for the quorum generator. It should take the random seed in the block header (always). I was wrong about it taking the N-1'th preimages, it's the N'th preimages.

Yes, that makes sense. And there is one more thing to be considered. Which pre-images are used for the `random seed` for the block header? There is a situation the pre-images are revealed for current height (I mean, the height of the currently externalized block) So, we need to generate the random seed with N-1'th preimages of the existing validator and N'th preimages of newly enrolled validators. What do you think of it?

Reply:
> There is a situation some pre-images are NOT revealed at the current height (I mean, the height of the currently externalized block)

I mean that's fine? After the block is externalized we use that random seed for the quorum generator. We don't care what's in it, just that it was voted and externalized.

Reply:
> I mean that's fine? After the block is externalized we use that random seed for the quorum generator. We don't care what's in it, just that it was voted and externalized.

The `random_seed` is not part of `ConsensusData`. So we should determine what pre-images should be in it, I think.

Reply:
I am just talking about the quorum generator though. Not the nomination process.

Reply:
If this isn't making much sense to you, it's fine to keep it the way you wrote it.

I'll submit a separate issue for this later.

Reply:
Sorry for the misunderstanding. I will think about what you mean again if you submit a new issue about that. Anyway, I just want to how to generate the `random_seed` for a block header.
Thank you.

Reply:
> Anyway, I just wanted to know how to generate the random_seed for a block header.

Yes, that's fair. I was looking at the other use-case and not specifically your own. So it may have been off-topic for this PR. I apologize.

Reply:
> > Anyway, I just wanted to know how to generate the random_seed for a block header.
> 
> Yes, that's fair. I was looking at the other use-case and not specifically your own. So it may have been off-topic for this PR. I apologize.

It's okay. There are some mistakes in expressing my opinion in English. Thanks.